### ISSUE TITLE
Extraction from HTML reports erroneous positions in some situations

### ISSUE THREAD

**Issue by davidtheclark** [stylelint/stylelint#1265]
This happens when the `<style>` content does not start with a newline, and when there are preceeding style tags with no newlines in them.

Here's some HTML that throws it off. I will post a PR with failing tests.

```html
<style>b { color: pink; }</style>
<style>a {}</style>
```

```html
<style
  id="horse"
  class="cow">a {}
</style>
```

```html
<style></style><style>a {}</style>
```

Also — while looking into these I noticed that the documentation for `htmlparser2` sucks. Is it the best HTML parsing library out there?

cc/ @blake-newman

**Comment by blake-newman** [comment]
@davidtheclark Not sure about the library. I've just used it in the past and so does eslint. 

example 1 & 3 i believe are somewhat fixable, by appending a new line before where there isn't and then accounting for that in the mappings. A little tricky but i'm sure it's possible, although i don't recommend styles that are 100% inline like shown, however we should try to support it if we can. 

The second example is a little more difficult. As the style tag itself is spanning across multiple lines, Im sure we can still resolve some how. This is most likely the least likely scenario and most difficult. So perhaps resolving this in another issue would mean quicker turnaround for time :)

**Comment by davidtheclark** [comment]
Whatever works!

I encountered these problems while looking into #1226. It seemed that that issue was pretty simple to solve by adding a newline to the end of the code. But when trying to test it I got flummoxed by these problems.

**Comment by davidtheclark** [comment]
@blake-newman: If it gets too ridiculous, we could always just say in the docs something like "We suggest that you begin and end each `<style>` tag's content with a newline. If you do not, those newlines will be added during processing and there might be inaccuracies in reported positions." Something to consider.

**Comment by evilebottnawi** [comment]
@davidtheclark why not calculate start index?

**Comment by davidtheclark** [comment]
@evilebottnawi If we can pull off always getting the correct position *and* not triggering rules erroneously, that's great! I was just proposing an option if we can't.

**Comment by evilebottnawi** [comment]
@davidtheclark i think this and https://github.com/stylelint/stylelint/issues/1226 difference issue, first we can close issue 1226 adding newline always, this issue i trying to help tomorrow or some late.

**Comment by davidtheclark** [comment]
Yes, #1226 can be addressed independently of the erroneous positions.

**Comment by blake-newman** [comment]
I'm fairly sure it's possible, without too much work. Ill take a look into it some point next week, if that's okay. Maybe for now add in docs, as a caveat and then when we resolve any issue we can remove.

**Comment by davidtheclark** [comment]
@blake-newman Any progress on this?

**Comment by blake-newman** [comment]
@davidtheclark 

Just came back from holiday, it's near the top of my priority list. So will be starting it, most scenarios should be fixable. Any that are not I'll just add some documentation. 

Once started shouldn't take me long. Sorry for the delay.

**Comment by blake-newman** [comment]
@davidtheclark I don't think i will get time at the moment to look into this. I've got a large project ongoing that is consuming most of my time.

**Comment by jeddy3** [comment]
Labelling as [processor](https://github.com/stylelint/stylelint/pull/1535) and closing.
